File: src/atcoder/abc311/abc311_d.scala
import java.util.Scanner
import scala.collection.mutable.Queue
import scala.collection.mutable.HashSet

object Main extends App {
  sealed trait Motion
  case object Up extends Motion
  case object Down extends Motion
  case object Left extends Motion
  case object Right extends Motion
  case object Stop extends Motion

  val sc = new Scanner(System.in)
  val h = sc.nextInt
  val w = sc.nextInt
  val g = Array.fill(h) { sc.next }
  var q = new Queue[(Int,Int,Motion)]()
  q += ((1, 1, Stop))

  var seen = Array.tabulate(h, w) { (y: Int, x: Int) => new HashSet[Motion]() }
  seen(1)(1) += Stop

  while (q.size > 0) {
    val (y, x, dir) = q.dequeue

    for {
      (dy, dx, ndir) <- Seq((-1, 0, Up), (1, 0, Down), (0, -1, Left), (0, 1, Right))
      if dir == Stop || dir == ndir
    } {
      val ny = y + dy
      val nx = x + dx

      if (g(ny)(nx) == '#') {
        if (dir != Stop) { q += ((y, x, Stop)) }
      } else if (!seen(ny)(nx).contains(ndir)) {
        seen(ny)(nx) += ndir
        q += ((ny, nx, ndir))
      }
    }
  }

  var ans = 0
  for (y <- 0 until h) {
    for (x <- 0 until w) {
      if (!seen(y)(x).isEmpty) {
        ans += 1
      }
    }
  }
  println(ans)
}

File: src/atcoder/abc311/abc311_d.swift
class Queue<T> {
  let capacity: Int
  private var buffer: [T?]
  private var lo = 0
  private var hi = -1

  public var count: Int {
    return (hi - lo + 1)
  }

  public init(capacity: Int) {
    self.capacity = capacity
    self.buffer = [T?](repeating: nil, count: capacity)
  }

  public func push(_ value: T) {
    hi += 1
    hi %= capacity
    buffer[hi] = value
  }

  public func shift() -> T? {
    let value = buffer[lo % capacity]
    lo += 1
    lo %= capacity
    return value
  }
}

enum State {
  case Up
  case Down
  case Left
  case Right
  case Stop
}

let line = readLine()!.split(separator: " ").map { Int($0)! }
let h = line[0]
let w = line[1]
var s = (0..<h).map { _ in [Character](readLine()!) }

var seen = (0..<h).map { _ in (0..<w).map { _ in Set<State>() }}
seen[1][1].insert(.Stop)

typealias Position = (Int, Int, State)

var q = Queue<Position>(capacity: h * w * 5)
q.push((1, 1, .Stop))

while q.count > 0 {
  let (y, x, dir) = q.shift()!

  for (dy, dx, ndir): Position in [
    (-1, 0, .Up),
    (1, 0, .Down),
    (0, 1, .Right),
    (0, -1, .Left)
  ] {
    if dir != .Stop && dir != ndir { continue }

    let ny = y + dy
    let nx = x + dx

    if s[ny][nx] == "#" {
      if dir != .Stop { q.push((y, x, .Stop)) }
      continue
    }

    if seen[ny][nx].contains(ndir) { continue }
    seen[ny][nx].insert(ndir)

    q.push((ny,nx,ndir))
  }
}

var ans = (0..<h).map { seen[$0].filter { !$0.isEmpty }.count }.reduce(0, +)
print(ans)
